wsmaster/codenvy-hosted-platform-api-impl/src/main/java/com/codenvy/api/workspace/LimitsCheckingWorkspaceManager.java

```diff
@@ -161,6 +186,35 @@ public WorkspaceImpl updateWorkspace(String id, Workspace update) throws Conflic
     }
 
     /**
+     * One of the checks in {@link #checkRamAndPropagateStart(WorkspaceConfig, String, String, String, WorkspaceCallback)}
+     * requests to docker for system ram amount. This request can take much time, so we can not mark the method as synchronized
+     * because this will reduce performance. The solution is to make the system ram check optimistic by limiting
+     * same time start workspace requests. All requests to start workspace performed at the same time can receive not actual
```

Comment:
> The solution is to make the system ram check optimistic by limitingsame time start workspace requests

This is not clear for me, can you please rephrase it.

As far as i understand, this doc should be based on a few important statements : 
- this method can't be performed more than N times and the semaphore is used to control that
- system ram check may be really slow as it is based on request to the swarm, so it may took some time for swarm to figure out all the memory allocated on nodes. 
-  the system ram check must be performed on each start method call as it mustn't be possible to start a new workspace if system ram limit exceeded
- the semaphore is a trade off between speed and risk to exceed system ram limit. In the worst case which is almost impossible you may get N(permits) workspaces started after the actually system limit was exceeded, while by acquiring an exclusive lock the system may stack in the state when swarm doesn't response and the lock is held by the thread that requests that information. On the other hand the start itself is asynchronous and pretty fast, so the throughput limit event a small one won't affect the performance.